usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_main.c

```diff
@@ -0,0 +1,419 @@
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2019 Nexenta Systems, Inc.
+ * Copyright 2019 RackTop Systems
+ */
+
+/*
+ * Driver attach/detach routines are found here.
+ */
+
+/* ---- Private header files ---- */
+#include <smartpqi.h>
+
+void	*pqi_state;
+
+/* ---- Autoconfigure forward declarations ---- */
+static int smartpqi_attach(dev_info_t *dip, ddi_attach_cmd_t cmd);
+static int smartpqi_detach(dev_info_t *dip, ddi_detach_cmd_t cmd);
+static int smartpqi_power(dev_info_t *dip, int component, int level);
+static int smartpqi_getinfo(dev_info_t *dip, ddi_info_cmd_t cmd, void *arg,
+    void **results);
+static int smartpqi_quiesce(dev_info_t *dip);
+
+/* ---- cb_ops forward declarations ---- */
+static int smartpqi_ioctl(dev_t dev, int cmd, intptr_t data, int mode,
+    cred_t *credp, int *rval);
+
+static struct cb_ops smartpqi_cb_ops = {
+	scsi_hba_open,		/* open */
+	scsi_hba_close,		/* close */
+	nodev,			/* strategy */
+	nodev,			/* print */
+	nodev,			/* dump */
+	nodev,			/* read */
+	nodev,			/* write */
+	smartpqi_ioctl,		/* ioctl */
+	nodev,			/* devmap */
+	nodev,			/* mmap */
+	nodev,			/* segmap */
+	nochpoll,		/* chpoll */
+	ddi_prop_op,		/* cb_prop_op */
+	NULL,			/* streamtab */
+	D_MP,			/* cb_flag */
+	CB_REV,			/* rev */
+	nodev,			/* aread */
+	nodev			/* awrite */
+};
+
+static struct dev_ops smartpqi_ops = {
+	DEVO_REV,		/* dev_rev */
+	0,			/* refcnt */
+	smartpqi_getinfo,	/* info */
+	nulldev,		/* identify */
+	nulldev,		/* probe */
+	smartpqi_attach,	/* attach */
+	smartpqi_detach,	/* detach */
+	nodev,			/* reset */
+	&smartpqi_cb_ops,	/* driver operations */
+	NULL,			/* bus operations */
+	smartpqi_power,		/* power management */
+	smartpqi_quiesce,	/* quiesce */
+};
+
+static struct modldrv modldrv = {
+	&mod_driverops,
+	SMARTPQI_MOD_STRING,
+	&smartpqi_ops,
+};
+
+static struct modlinkage modlinkage = {
+	MODREV_1, &modldrv, NULL
+};
+
+int pqi_do_scan = 0;
+int pqi_do_ctrl = 0;
+int pqi_offline_target = 0;
+int pqi_do_offline = 0;
+
+/*
+ * This is used for data I/O DMA memory allocation. (full 64-bit DMA
+ * physical addresses are supported.)
+ */
+ddi_dma_attr_t smartpqi_dma_attrs = {
+	DMA_ATTR_V0,		/* attribute layout version		*/
+	0x0ull,			/* address low - should be 0 (longlong)	*/
+	0xffffffffffffffffull, /* address high - 64-bit max	*/
+	0x00666600ull,		/* count max - max DMA object size	*/
+	4096,			/* allocation alignment requirements	*/
+	0x78,			/* burstsizes - binary encoded values	*/
+	1,			/* minxfer - gran. of DMA engine	*/
+	0x00666600ull,		/* maxxfer - gran. of DMA engine	*/
+	0x00666600ull,		/* max segment size (DMA boundary)	*/
```

Comment:
This value for the maximum segment size is almost *certainly* incorrect.  In general most modern devices probably can use either 2^32-1, or 2^64-1 here.  (Probably the latter.)  This field was intended for legacy DMA engines that might only have a subset of bits in their address lines -- typically legacy ISA devices and such which might restrict DMA to a 64k or 1MB segment at a time.